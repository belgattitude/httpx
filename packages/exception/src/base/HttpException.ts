import { supportsErrorCause } from '../support';
import type { HttpExceptionParams, HttpMethod } from '../types';
import { getSuper } from '../utils';

export class HttpException extends Error {
  /**
   * Http error status code (400-599)
   */
  public readonly statusCode: number;
  /**
   * Indicates the original url that caused the error.
   */
  public readonly url: string | undefined;

  /**
   * Http method
   */
  public readonly method: HttpMethod | undefined;

  /**
   * Custom additional code (ie: 'AbortError', 'CODE-1234'...)
   */
  public readonly code: string | undefined;

  /**
   * Inform about an unique error identifier (ie: nanoid, cuid...)
   */
  public readonly errorId: string | undefined;

  /**
   * If set and the runtime (browser or node) supports it
   * you can get back the error cause
   * @see https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Error/cause
   */
  public readonly cause?: Error | HttpException;

  /**
   * Construct a new HttpException class
   *
   * @param statusCode http status code between 400-599, no checks are done on the validity of the number.
   * @param msgOrParams either a message or an object containing HttpExceptionParams
   */
  constructor(statusCode: number, msgOrParams?: HttpExceptionParams | string) {
    const name = 'HttpException';
    const { message, url, cause, errorId, code, method } = getSuper(
      name,
      msgOrParams
    );
    super(message);
    if (supportsErrorCause() && cause instanceof Error) {
      this.cause = cause;
    }
    this.statusCode = statusCode;
    this.url = url;
    this.errorId = errorId;
    this.code = code;
    this.method = method;

    Object.setPrototypeOf(this, HttpException.prototype);
    this.name = name;
  }
}
