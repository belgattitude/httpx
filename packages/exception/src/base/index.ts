export { HttpException } from './HttpException';
export { HttpServerException } from './HttpServerException';
export { HttpClientException } from './HttpClientException';
