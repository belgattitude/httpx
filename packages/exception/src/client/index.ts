export { HttpBadRequest } from './HttpBadRequest';
export { HttpConflict } from './HttpConflict';
export { HttpExpectationFailed } from './HttpExpectationFailed';
export { HttpFailedDependency } from './HttpFailedDependency';
export { HttpForbidden } from './HttpForbidden';
export { HttpGone } from './HttpGone';
export { HttpImATeapot } from './HttpImATeapot';
export { HttpLengthRequired } from './HttpLengthRequired';
export { HttpLocked } from './HttpLocked';
export { HttpMisdirectedRequest } from './HttpMisdirectedRequest';
export { HttpMethodNotAllowed } from './HttpMethodNotAllowed';
export { HttpNotAcceptable } from './HttpNotAcceptable';
export { HttpNotFound } from './HttpNotFound';
export { HttpPayloadTooLarge } from './HttpPayloadTooLarge';
export { HttpPaymentRequired } from './HttpPaymentRequired';
export { HttpPreconditionFailed } from './HttpPreconditionFailed';
export { HttpPreconditionRequired } from './HttpPreconditionRequired';
export { HttpProxyAuthenticationRequired } from './HttpProxyAuthenticationRequired';
export { HttpRangeNotSatisfiable } from './HttpRangeNotSatisfiable';
export { HttpRequestHeaderFieldsTooLarge } from './HttpRequestHeaderFieldsTooLarge';
export { HttpRequestTimeout } from './HttpRequestTimeout';
export { HttpTooEarly } from './HttpTooEarly';
export { HttpTooManyRequests } from './HttpTooManyRequests';
export { HttpUnauthorized } from './HttpUnauthorized';
export { HttpUnavailableForLegalReasons } from './HttpUnavailableForLegalReasons';
export { HttpUnprocessableEntity } from './HttpUnprocessableEntity';
export { HttpUnsupportedMediaType } from './HttpUnsupportedMediaType';
export { HttpUpgradeRequired } from './HttpUpgradeRequired';
export { HttpUriTooLong } from './HttpUriTooLong';
