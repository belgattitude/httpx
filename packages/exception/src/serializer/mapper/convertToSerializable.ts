import type { HttpException } from '../../base';
import { isHttpException } from '../../typeguards';
import { isNativeError } from '../typeguard';
import type { NativeError, Serializable } from '../types';

/**
 * Convert an Error, NativeError or any HttpException to
 * an object suitable for serialization (a serializable version).
 *
 * @link {createFromSerializable}
 */
export const convertToSerializable = (
  e: Error | NativeError | HttpException
  // eslint-disable-next-line sonarjs/cognitive-complexity
): Serializable => {
  const {
    name,
    message,
    stack = null,
    cause: c = null,
  } = (
    (e as unknown) instanceof Error
      ? e
      : {
          name: 'Error',
          message:
            typeof (e as unknown) === 'string'
              ? e
              : `Can't serialize error at runtime. Received '${typeof e}'`,
        }
  ) as Error;
  const cause = c instanceof Error ? convertToSerializable(c) : null;
  const common = {
    name,
    message,
    ...(stack ? { stack } : {}),
    ...(cause ? { cause } : {}),
  };
  if (isHttpException(e)) {
    return {
      __type: 'HttpException',
      ...common,
      statusCode: e.statusCode,
      ...(e.url ? { url: e.url } : {}),
      ...(e.code ? { code: e.code } : {}),
      ...(e.method ? { method: e.method } : {}),
      ...(e.errorId ? { errorId: e.errorId } : {}),
    };
  }
  return {
    __type: isNativeError(e) ? 'NativeError' : 'NonNativeError',
    ...common,
  };
};
