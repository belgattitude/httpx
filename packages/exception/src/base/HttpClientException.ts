import type { HttpExceptionParams } from '../types';
import { getSuper } from '../utils';
import { HttpException } from './HttpException';

/**
 * Construct a new HttpClientException class
 *
 * @param statusCode http status code between 400-499, no checks are done on the validity of the number.
 * @param msgOrParams either a message or an object containing HttpExceptionParams
 */
export class HttpClientException extends HttpException {
  constructor(statusCode: number, msgOrParams?: HttpExceptionParams | string) {
    const name = 'HttpClientException';
    super(statusCode, getSuper(name, msgOrParams));
    Object.setPrototypeOf(this, HttpClientException.prototype);
    this.name = name;
  }
}
