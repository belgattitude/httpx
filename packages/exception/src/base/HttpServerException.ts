import type { HttpExceptionParams } from '../types';
import { getSuper } from '../utils';
import { HttpException } from './HttpException';

/**
 * Construct a new HttpServerException class
 *
 * @param statusCode http status code between 500-599, no checks are done on the validity of the number.
 * @param msgOrParams either a message or an object containing HttpExceptionParams
 */
export class HttpServerException extends HttpException {
  constructor(statusCode: number, msgOrParams?: HttpExceptionParams | string) {
    super(statusCode, getSuper('HttpServerException', msgOrParams));
    Object.setPrototypeOf(this, HttpServerException.prototype);
    this.name = 'HttpServerException';
  }
}
