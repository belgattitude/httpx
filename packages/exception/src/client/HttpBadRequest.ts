import { HttpClientException } from '../base';
import type { HttpExceptionParams } from '../types';
import { getSuper } from '../utils';

/**
 * 400 Bad Request (client)
 *
 * The server cannot or will not process the request due to something that is perceived to be a client error
 * (e.g., malformed request syntax, invalid request message framing, or deceptive request routing).
 *
 * @see https://httpstatus.in/400/
 * @see https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/400
 */
export class HttpBadRequest extends HttpClientException {
  static readonly STATUS = 400;
  constructor(msgOrParams?: HttpExceptionParams | string) {
    const name = 'BadRequest';
    super(400, getSuper(name, msgOrParams));
    Object.setPrototypeOf(this, HttpBadRequest.prototype);
    this.name = `Http${name}`;
  }
}
