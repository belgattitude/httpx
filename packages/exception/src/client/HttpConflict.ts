import { HttpClientException } from '../base';
import type { HttpExceptionParams } from '../types';
import { getSuper } from '../utils';

/**
 * 409 Conflict (client)
 *
 * This response is sent when a request conflicts with the current state of the server.
 *
 * @see https://httpstatus.in/409/
 * @see https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/409
 */
export class HttpConflict extends HttpClientException {
  static readonly STATUS = 409;
  constructor(msgOrParams?: HttpExceptionParams | string) {
    const name = 'Conflict';
    super(409, getSuper(name, msgOrParams));
    Object.setPrototypeOf(this, HttpConflict.prototype);
    this.name = `Http${name}`;
  }
}
