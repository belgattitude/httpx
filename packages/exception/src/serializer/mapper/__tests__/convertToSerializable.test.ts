import { HttpException } from '../../../base';
import { HttpBadRequest } from '../../../client';
import type { SerializableHttpException } from '../../types';
import { convertToSerializable } from '../convertToSerializable';

describe('convertToSerializable', () => {
  describe('when an http exception is given', () => {
    const httpExceptions: [label: string, error: HttpException][] = [
      ['simple', new HttpException(500)],
      ['withMsg', new HttpException(500, 'msg')],
      [
        'withFullParams',
        new HttpException(500, {
          errorId: 'nanoid()',
          message: 'msg',
          code: 'NETWORK_UNAVAILABLE',
          url: 'http://localhost',
          method: 'PUT',
        }),
      ],
      [
        'withCause',
        new HttpException(500, {
          cause: new HttpException(500, 'msg'),
        }),
      ],
      [
        'HttpBadRequest',
        new HttpBadRequest({
          message: 'msg',
          cause: new EvalError(),
          url: 'http://',
        }),
      ],
    ];

    it.each(httpExceptions)('works with %s', (label, err) => {
      const serializable = convertToSerializable(
        err
      ) as SerializableHttpException;
      expect(serializable.__type).toStrictEqual('HttpException');
      expect(serializable.message).toStrictEqual(err.message);
      expect(serializable.name).toStrictEqual(err.name);
      expect(serializable?.url).toStrictEqual(err?.url);
      expect(serializable?.code).toStrictEqual(err?.code);
      expect(serializable?.errorId).toStrictEqual(err?.errorId);
      expect(serializable?.method).toStrictEqual(err?.method);
      const { cause, stack, ...serializableWithoutCause } = serializable;
      expect(serializableWithoutCause).toMatchSnapshot();
    });
  });

  describe('when native error is given', () => {
    const natives: [label: string, err: Error][] = [
      ['ErrorSimple', new Error()],
      ['ErrorSimpleRange', new RangeError()],
      ['ErrorSimpleEval', new EvalError()],
      ['ErrorMsg', new Error('msg')],
      [
        'ErrorMsg',
        new Error('msg', {
          cause: new Error('msg'),
        }),
      ],
    ];
    it.each(natives)('works with %s', (label, err) => {
      const serializable = convertToSerializable(err);
      expect(serializable.__type).toStrictEqual('NativeError');
      expect(serializable.message).toStrictEqual(err.message);
      expect(serializable.name).toStrictEqual(err.name);
      const { cause, stack, ...serializableWithoutCause } = serializable;
      expect(serializableWithoutCause).toMatchSnapshot();
    });
  });

  describe('when non native error is given', () => {
    class NonNativeError extends Error {
      constructor(message: string) {
        super(message);
        Object.setPrototypeOf(this, NonNativeError.prototype);
        this.name = 'NonNativeError';
      }
    }
    it('set the type to NonNativeError', () => {
      const e = new NonNativeError('msg');
      const serializable = convertToSerializable(e);
      expect(e.name).toStrictEqual('NonNativeError');
      expect(serializable.__type).toStrictEqual('NonNativeError');
      const { cause, stack, ...serializableWithoutCause } = serializable;
      expect(serializableWithoutCause).toMatchSnapshot();
    });
  });

  describe('when a non error is given', () => {
    it('switch to build-in Error with custom message', () => {
      const e = new Date() as unknown as Error;
      const serializable = convertToSerializable(e);
      expect(serializable.__type).toStrictEqual('NonNativeError');
      const { cause, stack, ...serializableWithoutCause } = serializable;
      expect(serializableWithoutCause).toMatchSnapshot();
    });
  });
});
