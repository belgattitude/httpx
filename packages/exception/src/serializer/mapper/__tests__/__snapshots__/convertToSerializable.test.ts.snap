// Vitest Snapshot v1

exports[`convertToSerializable > when a non error is given > switch to build-in Error with custom message 1`] = `
{
  "__type": "NonNativeError",
  "message": "Can't serialize error at runtime. Received 'object'",
  "name": "Error",
}
`;

exports[`convertToSerializable > when an http exception is given > works with HttpBadRequest 1`] = `
{
  "__type": "HttpException",
  "message": "msg",
  "name": "HttpBadRequest",
  "statusCode": 400,
  "url": "http://",
}
`;

exports[`convertToSerializable > when an http exception is given > works with simple 1`] = `
{
  "__type": "HttpException",
  "message": "Http exception",
  "name": "HttpException",
  "statusCode": 500,
}
`;

exports[`convertToSerializable > when an http exception is given > works with withCause 1`] = `
{
  "__type": "HttpException",
  "message": "Http exception",
  "name": "HttpException",
  "statusCode": 500,
}
`;

exports[`convertToSerializable > when an http exception is given > works with withFullParams 1`] = `
{
  "__type": "HttpException",
  "code": "NETWORK_UNAVAILABLE",
  "errorId": "nanoid()",
  "message": "msg",
  "method": "PUT",
  "name": "HttpException",
  "statusCode": 500,
  "url": "http://localhost",
}
`;

exports[`convertToSerializable > when an http exception is given > works with withMsg 1`] = `
{
  "__type": "HttpException",
  "message": "msg",
  "name": "HttpException",
  "statusCode": 500,
}
`;

exports[`convertToSerializable > when native error is given > works with ErrorMsg 1`] = `
{
  "__type": "NativeError",
  "message": "msg",
  "name": "Error",
}
`;

exports[`convertToSerializable > when native error is given > works with ErrorMsg 2`] = `
{
  "__type": "NativeError",
  "message": "msg",
  "name": "Error",
}
`;

exports[`convertToSerializable > when native error is given > works with ErrorSimple 1`] = `
{
  "__type": "NativeError",
  "message": "",
  "name": "Error",
}
`;

exports[`convertToSerializable > when native error is given > works with ErrorSimpleEval 1`] = `
{
  "__type": "NativeError",
  "message": "",
  "name": "EvalError",
}
`;

exports[`convertToSerializable > when native error is given > works with ErrorSimpleRange 1`] = `
{
  "__type": "NativeError",
  "message": "",
  "name": "RangeError",
}
`;

exports[`convertToSerializable > when non native error is given > set the type to NonNativeError 1`] = `
{
  "__type": "NonNativeError",
  "message": "msg",
  "name": "NonNativeError",
}
`;
